conformance/utils/kubernetes/apply.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"io/ioutil"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// MustApplyWithCleanup creates or updates Kubernetes resources defined with the
+// provided YAML file and registers a cleanup function for resources it created.
+// Note that this does not remove resources that already existed in the cluster.
+func MustApplyWithCleanup(t *testing.T, c client.Client, path string, gcName string, cleanup bool) {
+	b, err := ioutil.ReadFile(path)
+	require.NoErrorf(t, err, "error reading %s file", path)
+
+	decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewReader(b), 4096)
+	for {
+		uObj := unstructured.Unstructured{}
+		if decodeErr := decoder.Decode(&uObj); decodeErr != nil {
+			if errors.Is(decodeErr, io.EOF) {
+				break
+			}
+			t.Logf("manifest: %s", string(b))
+			require.NoErrorf(t, decodeErr, "error parsing manifest")
+		}
+		if len(uObj.Object) == 0 {
+			continue
+		}
+
+		if uObj.GetKind() == "Gateway" {
+			err = unstructured.SetNestedField(uObj.Object, gcName, "spec", "gatewayClassName")
+			require.NoErrorf(t, err, "error setting GatewayClass")
+		}
+		namespacedName := types.NamespacedName{Namespace: uObj.GetNamespace(), Name: uObj.GetName()}
+		fetchedObj := uObj.DeepCopy()
+		err := c.Get(context.TODO(), namespacedName, fetchedObj)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				require.NoErrorf(t, err, "error getting resource")
+			}
+			t.Logf("Creating %s %s", uObj.GetName(), uObj.GetKind())
```

Comment:
Add a similar log line before c.Get() on line 62.

Reply:
I'm a little hesitant to do that because our test output is already fairly verbose here. Here's the output I'm currently getting:

```
    suite.go:56: Test Setup: Ensuring GatewayClass has been accepted
    suite.go:59: Test Setup: Applying base manifests
    apply.go:81: Updating gateway-conformance-infra Namespace
    apply.go:81: Updating same-namespace Gateway
    apply.go:81: Updating all-namespaces Gateway
    apply.go:81: Updating backend-namespaces Gateway
    apply.go:81: Updating infra-backend-v1 Service
    apply.go:81: Updating infra-backend-v1 Deployment
    apply.go:81: Updating infra-backend-v2 Service
    apply.go:81: Updating infra-backend-v2 Deployment
    apply.go:81: Updating gateway-conformance-app-backend Namespace
    apply.go:81: Updating app-backend-v1 Service
    apply.go:81: Updating app-backend-v1 Deployment
    apply.go:81: Updating app-backend-v2 Service
    apply.go:81: Updating app-backend-v2 Deployment
    apply.go:81: Updating gateway-conformance-web-backend Namespace
    apply.go:81: Updating web-backend Service
    apply.go:81: Updating web-backend Deployment
    suite.go:62: Test Setup: Ensuring Gateways and Pods from base manifests are ready
```

Sometimes there's a mix of `Updating` and `Creating` depending on if I've enabled cleanup or added new resources to the base manifests. In either case, adding a `Getting` before each `Updating` or `Creating` line may add more confusion or noise than it's worth. I think it could be helpful to log before the first API call we're doing in case of a timeout or other connection issue, but that's actually happening earlier in the test flow. Do you think there's a better way to log this output as a whole?